server/src/api/requests/quantity.ts

```diff
@@ -0,0 +1,119 @@
+import {RequestStatus} from "../graphql.types";
+import {DB} from "../../database";
+import {Item} from "../items/item";
+
+interface QuantitiesInStatus {
+    [statusName: string]: number;
+}
+
+interface ItemQuantities {
+    [itemId: string]: QuantitiesInStatus;
+}
+
+interface ItemQtyAvailable {
+    [itemId: string]: number;
+}
+
+export class Quantity {
+    public static async inStock(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        return this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["FULFILLED", "LOST", "DAMAGED"], itemId);
+    }
+
+    public static async unreserved(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        return this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+    }
+
+    public static async availableForApproval(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        return this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+    }
+
+    public static async all(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        const qtyInStock: ItemQtyAvailable = this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["FULFILLED", "LOST", "DAMAGED"], itemId);
+        const qtyUnreserved: ItemQtyAvailable = this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+        const qtyAvailableForApproval: ItemQtyAvailable = this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+
```

Comment:
I'm guessing this is the endpoint that's gonna get hit by client all the time, with websockets maybe consider leaving a TODO to do this in one compact calculation rather than three calls

Reply:
It's 3 separate calls but the expensive part (DB queries) only get called once.